[BlazorWebView] Fixed Allow force reload for URLs with dots in query parameters

tw4 opened the pull request:
<!-- Please let the below note in for people that find this PR -->
> [!NOTE]  
> Are you waiting for the changes in this PR to be merged?  
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

<!--
!!!!!!! MAIN IS THE ONLY ACTIVE BRANCH. MAKE SURE THIS PR IS TARGETING MAIN. !!!!!!!
-->

### Description of Change

This PR fixes a bug in `BlazorWebView` where URLs containing dots (`.`) in **query parameters** (e.g., `?weight=62.5`) could not be force reloaded. The issue was caused by using `Path.HasExtension(uriString)`, which mistakenly treated dots in query strings as if they were part of a file extension.

---

#### ❌ Before

The original logic:

```csharp
if (Path.HasExtension(uriString))
    return false;
```

This evaluates the entire URI string — including query parameters — which incorrectly caused `?value=62.5` to be treated as a file reference, breaking reload behavior.

---

#### ✅ After

This PR introduces a safer and more accurate approach:

```csharp
if (Path.HasExtension(uri.AbsolutePath))
    return false;
```

Along with additional improvements:

* Validates input with `string.IsNullOrWhiteSpace`
* Uses `Uri.TryCreate` with `UriKind.Absolute` to safely parse the input and reject relative or malformed URIs
* Scheme/host matching is preserved through `Uri.IsBaseOf`, so URIs whose scheme or host does not match the configured base URI are still rejected as before

This ensures that only the actual path is evaluated for file extensions, allowing query strings to contain dots without affecting navigation or reload logic.

---

### Issues Fixed

Fixes #25689

Cover test result
<img width="477" alt="Screenshot 2025-05-16 at 22 54 41" src="https://github.com/user-attachments/assets/55e5948b-17e6-4abd-9fad-d01172855c43" />


---

**Note:** @mattleibow  
The previous PR might have been truly cursed 😅 — I created a new branch, rewrote the unit test project from scratch, and everything worked like a charm this time.  
No issues so far, and I'm optimistic GitHub Actions will behave too! 🧙‍♂️✨

Also, big thanks to @mattleibow for the support and guidance — your insights helped me see things more clearly! 🙏

tw4 committed:
fix: Allow force reload for URLs with dots in query parameters

jsuarezruiz commented:
/azp run

jsuarezruiz commented on src/BlazorWebView/tests/MauiBlazorWebView.UnitTests/Extensions_Tests.cs:
```diff
@@ -0,0 +1,82 @@
+	[InlineData("https://example.com/test", true)]
+	[InlineData("https://example.com/folder/subfolder/", true)]
+	[InlineData("https://example.com/folder/file.exe", false)]
+	public void IsBaseOfPage_HandlesVariousUris(string? uriString, bool expected)
```
Can include some additional cases?

```
[InlineData("https://subdomain.example.com/page", true)]  // Subdomain
[InlineData("https://example.com/path/with/.dot/segment", true)]  // Special character in path
[InlineData("https://example.com/path/with space", true)]  // Space in path
[InlineData("https://example.com/path/with%20encoded%20space", true)]  // URL encoded path
[InlineData("https://example.com/page?param=value&param2=value2", true)]  // Multiple query parameters
[InlineData("https://example.com/page.html", false)]  // HTML file
[InlineData("HTTPS://EXAMPLE.COM/PAGE", true)]  // Uppercase scheme and host
```

tw4 replied:
Thank you for your suggestions! I have added the test cases you mentioned, and I noticed that my code was not handling subdomains correctly. After reviewing the code and extending the tests to include more subdomain scenarios, the implementation now also passes all subdomain-related test cases.

tw4 replied:
I added the extra test cases (path with a . segment, space/encoded space, multiple query params, .html, uppercase scheme/host). I kept Uri.IsBaseOf semantics for subdomains; since the host differs from baseUri, the expectation is false. If subdomains should be treated as base, I can update that.

tw4 committed:
Fix: Add support for subdomain URLs in IsBaseOfPage and expand test coverage

bhavanesh2001 commented on src/BlazorWebView/src/Maui/Extensions/UriExtensions.cs:
```diff
@@ -1,20 +1,30 @@
-				// If the path ends in a file extension, it's not referring to a page.
-				return false;
-			}
+		if (string.IsNullOrWhiteSpace(uriString))
+			return false;
```
I think people here prefer curly braces. There are a few other places.
```suggestion
		if (string.IsNullOrWhiteSpace(uriString))
		{
			return false;
		}
```

tw4 replied:
Thank you for the feedback.

tw4 committed:
style: use curly braces for single-line statements

kubaflo reviewed (changes requested):
Could you please review the AI's summary?

tw4 committed:
fix: address AI review feedback - restore Uri.IsBaseOf semantics

- Remove custom subdomain host/path matching, preserve original
  Uri.IsBaseOf behavior as the scope of this fix is limited to
  query/fragment dot handling
- Use uri.AbsolutePath for Path.HasExtension check (core bug fix)
- Align test project with repo conventions: use $(_MauiDotNetTfm)
  and centralized package version variables
- Update subdomain test expectations to false (Uri.IsBaseOf does
  not match across different hosts)

tw4 committed:
Move MauiBlazorWebView.UnitTests into main solutions and fix test name

- Add MauiBlazorWebView.UnitTests project to Microsoft.Maui.sln,
  Microsoft.Maui-dev.sln, Microsoft.Maui-vscode.sln,
  Microsoft.Maui.LegacyControlGallery.sln, Microsoft.Maui-mac.slnf,
  and Microsoft.Maui-windows.slnf
- Remove standalone MauiBlazorWebView.sln (no longer needed)
- Rename IsBaseOfPage_IgnoresDotInQuery_AndReturnsFalse_WhenNotBase
  to IsBaseOfPage_IgnoresDotInQuery_WhenBaseUriMatches to reflect
  the actual expected result
- Update test namespace to Microsoft.Maui.MauiBlazorWebView.UnitTests
  to match InternalsVisibleTo attribute

tw4 committed:
Merge remote-tracking branch 'upstream/main' into fix-25689

# Conflicts:
#	Microsoft.Maui-mac.slnf
#	Microsoft.Maui-windows.slnf

tw4 committed:
Add MauiBlazorWebView.UnitTests to dotnet-test and Helix test lists

The new project was included in the solution files but was not enumerated
in eng/cake/dotnet.cake or eng/helix.proj, so the IsBaseOfPage regression
tests would build but never run in CI.

tw4 committed:
Rename scheme-mismatch test to reflect actual behavior

The test name implied an explicit http/https scheme guard, but
IsBaseOfPage relies on Uri.IsBaseOf for scheme/host matching.

tw4 committed:
Use GetComponents for extension check to handle percent-encoded paths

kubaflo reviewed (changes requested):
Could you check the ai's summary?

tw4 committed:
style: rename Extensions_Tests.cs to UriExtensions_Tests.cs to match class name

kubaflo reviewed (changes requested):
Could you please check the ai's suggestions?

tw4 committed:
Test IsBaseOfPage against real platform app origins

The regression tests only exercised a generic https://example.com host,
but BlazorWebView uses https://0.0.0.1/ (Android/Windows) and app://0.0.0.1/
(iOS/MacCatalyst) as the app origin. Add a theory covering dot-in-query,
dot-in-fragment and real-extension cases for those actual origins so the
issue #25689 scenario is reproduced on the platforms it affects.

kubaflo commented:
/review -b feature/enhanced-reviewer
